homeassistant/components/moon/sensor.py

```diff
@@ -86,4 +102,35 @@ def icon(self):
     async def async_update(self):
         """Get the time and updates the states."""
         today = dt_util.as_local(dt_util.utcnow()).date()
+        self._calculate_next_moon_dates()
         self._state = Astral().moon_phase(today)
+
+    def _calculate_next_moon_dates(self):
+        # if we already have dates that are in the future, skip the calculation
+        today = dt_util.as_local(dt_util.utcnow()).date()
+        if (
+            self.next_new_moon is not None
+            and self.next_full_moon is not None
+            and self.next_new_moon > today
+            and self.next_full_moon > today
+        ):
+            return
+
+        # Astral doesn't provide the next dates, so let's iterate the
+        # next 31 days (full moon phase lasts about 29,5 days)
+        # to find the next new and full moon days
+        astral = Astral()
+        end = today + timedelta(days=31)
+        new_moon = None
+        full_moon = None
+        for i in range((end - today).days):
+            day = today + timedelta(days=i)
+            phase = round(astral.moon_phase(date=day, rtype=float), 0)
+            if phase == 0:
+                new_moon = day
+            if phase == 14:
+                full_moon = day
+            if new_moon is not None and full_moon is not None:
+                break
```

Comment:
IMHO this should be moved to `astral` since it is not HA specific.

Reply:
Yes, that should be part of `astral`.

Reply:
Therefore, please open a PR upstream to update `astral`, thanks 👍

Reply:
Done. Waiting for merge...

Reply:
Great! Could you link the PR so we can track it as well? Thanks :)

Reply:
https://github.com/sffjunkie/astral/pull/51

Reply:
So, there is no reaction on the MR in the upstream. What next?

Reply:
We can continue here when the upstream PR is merged.